File: addons/webview_mobile/android/src/main/kotlin/com/al67x/webview/AL67XWebViewPlugin.kt
package com.al67x.webview

import android.webkit.JavascriptInterface
import android.webkit.WebView
import android.webkit.WebViewClient
import org.godotengine.godot.Godot
import org.godotengine.godot.plugin.GodotPlugin
import org.godotengine.godot.plugin.SignalInfo
import java.io.File

/**
 * AL67XWebViewPlugin — Android backend for WebViewHost (see
 * src/services/webview_host.gd + addons/webview_mobile/README.md).
 *
 * STATUS: source skeleton only, not compiled/tested in this repo (no
 * Android SDK/NDK in the dev sandbox this was written in). Wraps the
 * platform's own android.webkit.WebView (Chromium-backed on modern
 * devices) per the "auto-detect the native device webview" request.
 *
 * Registers as an Engine singleton named "AL67XWebView" exposing:
 *   load_html_file(res_path: String)
 *   post_message(json: String)
 *   signal ipc_message(message: String)
 * matching exactly what WebViewHost._mount_native_mobile() expects.
 */
class AL67XWebViewPlugin(godot: Godot) : GodotPlugin(godot) {

    private var webView: WebView? = null

    override fun getPluginName(): String = "AL67XWebView"

    override fun getPluginSignals(): Set<SignalInfo> =
        setOf(SignalInfo("ipc_message", String::class.java))

    /** Called from GDScript: AL67XWebView.load_html_file("res://web_ui/hub/index.html") */
    fun load_html_file(resPath: String) {
        runOnUiThread {
            ensureWebView()
            // res:// paths must already be resolved to a real filesystem path
            // by the caller (Godot's ProjectSettings.globalize_path) before
            // reaching here, or exposed via a local content:// bridge —
            // whichever this plugin's Kotlin-side companion helper ends up
            // using. Left as a TODO since it depends on how assets are
            // packaged (APK asset:// vs extracted user:// data dir).
            val fileUrl = "file://" + File(resPath).absolutePath
            webView?.loadUrl(fileUrl)
        }
    }

    fun post_message(json: String) {
        runOnUiThread {
            val escaped = json.replace("\\", "\\\\").replace("'", "\\'")
            webView?.evaluateJavascript(
                "document.dispatchEvent(new CustomEvent('og-message', { detail: '$escaped' }));" +
                    "document.dispatchEvent(new CustomEvent('message', { detail: '$escaped' }));",
                null
            )
        }
    }

    private fun ensureWebView() {
        if (webView != null) return
        val activity = godot.requireActivity()
        val wv = WebView(activity)
        wv.settings.javaScriptEnabled = true
        wv.settings.domStorageEnabled = true
        wv.webViewClient = WebViewClient()
        wv.addJavascriptInterface(IpcBridge(), "ipc")
        webView = wv
        // TODO: actually attach `wv` into the Godot view hierarchy at the
        // WebViewHost Control's screen-space rect (this skeleton doesn't
        // wire layout/sizing yet — needs the same "sync rect on resize"
        // treatment WebViewHost's Web-export path already does).
    }

    /** Exposed to JS as `window.ipc` -> ipc.postMessage(json) */
    inner class IpcBridge {
        @JavascriptInterface
        fun postMessage(message: String) {
            emitSignal("ipc_message", message)
        }
    }
}

File: addons/webview_mobile/ios/AL67XWebViewPlugin.swift
import Foundation
import WebKit

/// AL67XWebViewPlugin — iOS backend for WebViewHost (see
/// src/services/webview_host.gd + addons/webview_mobile/README.md).
///
/// STATUS: source skeleton only, not compiled/tested in this repo (no
/// Xcode/macOS toolchain in the dev sandbox this was written in). Wraps
/// WKWebView — literally Safari's engine — per the "auto-detect the
/// native device webview" request.
///
/// Registers as an Engine singleton named "AL67XWebView" exposing:
///   load_html_file(res_path: String)
///   post_message(json: String)
///   signal ipc_message(message: String)
/// matching exactly what WebViewHost._mount_native_mobile() expects.
///
/// This file is written against Godot's documented iOS plugin shape
/// (a GodotPlugin-style Objective-C++/Swift bridge target); the exact
/// base class/registration macro names depend on the Godot version's iOS
/// plugin template — verify against
/// https://docs.godotengine.org/en/stable/tutorials/platform/ios/ios_plugin.html
/// when wiring this into a real Xcode project.
@objc(AL67XWebViewPlugin)
class AL67XWebViewPlugin: NSObject, WKScriptMessageHandler {

    static let pluginName = "AL67XWebView"

    private var webView: WKWebView?

    /// Called from GDScript: AL67XWebView.load_html_file("res://web_ui/hub/index.html")
    @objc func load_html_file(_ resPath: String) {
        DispatchQueue.main.async {
            self.ensureWebView()
            // TODO: resolve the res:// path to the bundled resource URL
            // (Godot's iOS export bundles res:// under the main bundle) and
            // call webView?.loadFileURL(...). Left unresolved since it
            // depends on the exact export-time resource layout.
            guard let url = Bundle.main.url(forResource: resPath, withExtension: nil) else { return }
            self.webView?.loadFileURL(url, allowingReadAccessTo: url.deletingLastPathComponent())
        }
    }

    @objc func post_message(_ json: String) {
        DispatchQueue.main.async {
            let escaped = json.replacingOccurrences(of: "\\", with: "\\\\")
                               .replacingOccurrences(of: "'", with: "\\'")
            let js = """
            document.dispatchEvent(new CustomEvent('og-message', { detail: '\(escaped)' }));
            document.dispatchEvent(new CustomEvent('message', { detail: '\(escaped)' }));
            """
            self.webView?.evaluateJavaScript(js, completionHandler: nil)
        }
    }

    private func ensureWebView() {
        if webView != nil { return }
        let config = WKWebViewConfiguration()
        config.userContentController.add(self, name: "ipc")
        let wv = WKWebView(frame: .zero, configuration: config)
        webView = wv
        // TODO: attach `wv` as a subview of the Godot root view controller,
        // positioned to match the WebViewHost Control's screen-space rect
        // (mirrors the rect-sync treatment WebViewHost's Web-export path
        // already does for its iframe overlay).
    }

    // WKScriptMessageHandler — JS calls window.webkit.messageHandlers.ipc.postMessage(json)
    func userContentController(_ userContentController: WKUserContentController,
                                didReceive message: WKScriptMessage) {
        guard message.name == "ipc", let body = message.body as? String else { return }
        // TODO: emit the "ipc_message" Godot signal via whichever
        // GodotPlugin signal-emission API this Godot version's iOS plugin
        // template exposes (varies by version — see the docs link above).
        NotificationCenter.default.post(name: Notification.Name("AL67XWebViewIpcMessage"), object: body)
    }
}

// JS-side note (mirrors web_ui/shared/bridge.js's transport-detection):
// WKWebView delivers JS->native via
//   window.webkit.messageHandlers.ipc.postMessage(json)
// not window.ipc.postMessage(). The native-mobile shim that
// WebViewHost/bridge.js expects (window.__OG_TRANSPORT__) should inject a
// small JS snippet on page load that aliases:
//   window.__OG_TRANSPORT__ = { postMessage: (m) => window.webkit.messageHandlers.ipc.postMessage(m) };
